// @7kprotocol/sdk-ts/lib/cjs/features/limitDca/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LO_DCA_API = exports.DCA_MAX_RATE = exports.CONFIG_MODULE_ID = exports.DCA_ORDER_MODULE_ID = exports.LIMIT_ORDER_MODULE_ID = exports.PACKAGE = exports.SLIPPAGE_SCALE = exports.RATE_SCALE = exports.GLOBAL_CONFIG_ID = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
exports.GLOBAL_CONFIG_ID = "0xffc7f420b481ed7a012b42e125fd7e5716b5673759cf4629e5fcfebce970a72d";
exports.RATE_SCALE = BigInt("1000000000000");
exports.SLIPPAGE_SCALE = BigInt("10000");
exports.PACKAGE = "0x0ef0f805710cf53c10f29c4f73e1144a662747e1839689a846e2520cae122adc";
exports.LIMIT_ORDER_MODULE_ID = `${exports.PACKAGE}::limit_order`;
exports.DCA_ORDER_MODULE_ID = `${exports.PACKAGE}::dca_order`;
exports.CONFIG_MODULE_ID = `${exports.PACKAGE}::config`;
exports.DCA_MAX_RATE = new bignumber_js_1.default("18446744073709551615");
exports.LO_DCA_API = "https://lod-dca.7k.ag/api";
