// @7kprotocol/sdk-ts/lib/cjs/libs/protocols/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtocolContract = void 0;
const flowx_1 = require("./flowx");
const aftermath_1 = require("./aftermath");
const cetus_1 = require("./cetus");
const deepbook_1 = require("./deepbook");
const kriya_1 = require("./kriya");
const turbos_1 = require("./turbos");
const suiswap_1 = require("./suiswap");
const bluemove_1 = require("./bluemove");
const kriyaV3_1 = require("./kriyaV3");
const sponsored_1 = require("./deepbookV3/sponsored");
const bluefin_1 = require("./bluefin");
const springsui_1 = require("./springsui");
const obric_1 = require("./obric");
const stsui_1 = require("./stsui");
const flowxV3_1 = require("./flowxV3");
const steamm_1 = require("./steamm");
const magma_1 = require("./magma");
exports.ProtocolContract = {
    cetus: cetus_1.CetusContract,
    turbos: turbos_1.TurbosContract,
    bluemove: bluemove_1.BluemoveContract,
    kriya: kriya_1.KriyaContract,
    suiswap: suiswap_1.SuiswapContract,
    aftermath: aftermath_1.AfterMathContract,
    deepbook: deepbook_1.DeepBookContract,
    deepbook_v3: sponsored_1.SponsoredDeepBookV3Contract,
    flowx: flowx_1.FlowXContract,
    flowx_v3: flowxV3_1.FlowxV3Contract,
    kriya_v3: kriyaV3_1.KriyaV3Contract,
    bluefin: bluefin_1.BluefinContract,
    springsui: springsui_1.SpringSuiContract,
    obric: obric_1.ObricContract,
    stsui: stsui_1.StSuiContract,
    steamm: steamm_1.SteammContract,
    magma: magma_1.MagmaContract,
};
